FLIP: General metadata (#2435)
google/flax: docs/flip/2434-general-metadata.md

# FLIP: Axis Metadata

- Start Date: 2022-08-08
- FLIP Issue: [#2434](https://github.com/google/flax/issues/2434)
- FLIP PR: [#2435](https://github.com/google/flax/pull/2435)
- Status: Proposal

## Summary

This FLIP proposes to extend Flax's variable collections with a generic axis metadata API. 
The core of the API is a abstract base class that is recognized by lifting transformations that can add an axis (vmap, scan).
Users can extend the base class to keep track of per-axis metadata in a way that works with lifted transformations.

## Motivation

Generally, there is no way in Flax to track metadata for variables across lifted transformations. Axis metadata is used to keep track of semantic information about axes into other (Flax independent) APIs.
For example, optimizers like AdaFactor can be configured on a per-axis level and partitioning APIs in JAX like xmap or pjit require per variable annotations to map effectiently to parallel hardware.

> Currently, there is experimental support for partitioning annotations which requires using dedicated wrapper around lifted transforms that change axes (``nn.scan``, ``nn.vmap``) and a special APIs to create variables (``param_with_axes`` and ``variable_with_axes``). 

Reviewer: > which requires using dedicated wrapper around lifted transforms that change axes

I don't understand this, could you clarify? Which "dedicated" wrapper are you talking about?

Author: Clarified

> The experimental partitioning API stores the metadata in a seperate collection named "[collection]_axes".

Reviewer: It would be useful to link to the API (I suppose everything is in `linen/partitioning.py`?

Author: done

The experimental API has a number of shortcomings that we like to solve:
1. The current API works for tracking PartitionSpecs but not for other types of metadata like optimizer annotations.
2. The implementation using an "xxx_axes" collection requires error-prone and non-composable string manipulation.
3. Special, partioning-aware variable creators and lifted transforms are required
4. The partioning API is hard to use with pre-existing Modules that aren't partioning aware.

## Proposal

To generalize metadata tracking and keep the specific metadata out of core Flax we propose the following abstract base class:

```python
class AxisMetadata(metaclass=abc.ABCMeta):

  @abc.abstractmethod
  def unbox(self) -> Any:
    pass

  @abc.abstractmethod

>   def add_axis(self, index: int, params: Dict[Any, Any]) -> TAxisMetadata:

Reviewer: What is not entirely clear to me is how `add_axis` and `remove_axis` will be implemented in classes implementing `AxisMetadata`. The way I understand it is that lifted transforms can call these functions to add additional axes where they can store data (for instance of a lifted scan). Are there other use cases for `add_axis`? Maybe you could give an example implementation to these functions just to make things even more clear?

Author: Added reference implementation in the Partitioned example below

    pass

  @abc.abstractmethod
  def remove_axis(self, index: int, params: Dict[Any, Any]) -> TAxisMetadata:
    pass
```

We call this type of class wrapping a value and keeping track of some additional data a **box**.
By defining an abstract base class for this box, the API does not need to be aware of the specifics of the metadata that is tracked.
This should make the API future proof and modular.

The ``add_axis`` and ``remove_axis`` callback return an instance of their own type instead of mutating in-place.
Typically, an implementation would be a ``flax.struct.PyTreeNode`` because the box should still be a valid JAX value and must therefore be handled by the PyTree API.
Calling ``jax.tree_map`` on a boxed value will simply map over the value in the box.
The lifted transforms that need to handle metadata will call ``jax.tree_map(..., is_leaf=lambda x: isinstance(x, AxisMetadata))`` to find the AxisMetadata instances within a PyTree.

Advantages of the boxing approach:

> 1. Boxing can be used outside of Flax and metadata is automatically "inherited". For example, the compiler state will

Reviewer: What do you mean by "the compiler state"?

Author: Should have been optimizer, fixed

   have the same partitioning spec as the parameters, because the state is initialized using a ``jax.tree_map`` over the boxed parameters.
2. Boxes are composable.
3. Boxing avoids string manipulation and generally avoids having to handle additional auxilary collections like "param_axes" in the current
   partitioning API.
4. No need to lift metadata collections seperately.

Disadvantages:
1. Handling boxed values requires the relatively new ``is_leaf=`` syntax which users might not be familiar with. Although users will

>    probably call Flax provided utils that handle teh low-level tree_map calls in most cases.

Reviewer: ```suggestion
   probably call Flax provided utils that handle the low-level tree_map calls in most cases.
```

3. Custom Pytree nodes have a small runtime overhead. It's hard to observe this in practise because JAX calls are async.

### Init syntax

Boxes can be created directly by the init function of a variable. Therefore, we propose to create metadata using higher-order initializers.

> The main advantage of this is that we can decouple metadata handling completly from the Module definition. Also, most Modules already over

Reviewer: ```suggestion
The main advantage of this is that we can decouple metadata handling compltely from the Module definition. Also, most Modules already overwrite
```

attributes to override the default initialzers so users can add metadata to existing Modules without requiring any code changes.

> Too illustrate this, let's consider a metadata class that keeps track of PartitionSpecs used by ``pjit``:

Reviewer: ```suggestion
To illustrate this, let's consider a metadata class that keeps track of PartitionSpecs used by ``pjit``:
```

```python
class Partitioned(flax.struct.PyTreeNode, AxisMetadata):
  value: Any

>   named_axes: Tuple[Optional[str]]

Reviewer: Nit: isn't this defining a tuple containing either `None` or a single `str`?

Author: Fixed

  
  ...

def with_partitioning(init_fn, names):
  def wrapper(*args, **kwargs):
    return Partitioned(init_fn(*args, **kwargs), names)
  return wrapper
```

Here we also defined a small utility called ``with_partitioning`` that we can use to wrap existing initialzers to add metadata: 

```python
# init kernel with lecun normal and split the output features over the data axis

> partitioned_dense = nn.Dense(features, kernel_init=with_partitioning(nn.initializers.lecun_normal, [None, "data"]))

Reviewer: ```suggestion
partitioned_dense = nn.Dense(features, kernel_init=with_partitioning(nn.initializers.lecun_normal, (None, "data")))
```

```

### Unbox syntax

> Metadata typically doesn't need to be handled by Modules directly. Therefore, we prosose to make Modules agnostic to Metadata boxes by default.

Reviewer: ```suggestion
Metadata typically doesn't need to be handled by Modules directly. Therefore, we propose to make Modules agnostic to Metadata boxes by default.
```

partitioned_dense = nn.Dense(features, kernel_init=with_partitioning(nn.initializers.lecun_normal, (None, "data")))
```

### Unbox syntax

Metadata typically doesn't need to be handled by Modules directly. Therefore, we propose to make Modules agnostic to Metadata boxes by default.
The ``unbox`` method can be used to unpack a variable such that only the original JAX arrays remain. Users can manually call unbox but to make
sure Module classes don't have to call it everywhere we add an unbox keyword arg to variable returning APIs (e.g.: ``.param``, ``.variable``, ``.get_variable``).
The keyword arg ``unbox`` will default to ``True`` such that a Modules are metadata agnostic by default. This also means existing Modules will be backward compatible
with the new API.

```python
kernel = self.param("kernel", self.kernel_init, shape)  # No AxisMetadata instances
kernel_box = self.get_variable("param", "kernel", unbox=False)  # AxisMetadata boxes are preserved
```

### Lift syntax

When calling a lifted transformation that adds an axis you will now be able to pass a dictionary with arguments.
These params will be passed to ``AxisMetadata`` add_axis/remove_axis callbacks:

```python

> nn.scan(..., variable_axes={"params": 0}, metadata_params={nn.Partitioned.AXIS_NAME: "layers"})

Reviewer 2: just making sure I follow - this is fed into the `params` argument of the add/remove axis routines - to help support the more generalized annotations that e.g. pax uses?

Author: exactly

# FLIP: Axis Metadata

- Start Date: 2022-08-08
- FLIP Issue: [#2434](https://github.com/google/flax/issues/2434)
- FLIP PR: [#2435](https://github.com/google/flax/pull/2435)
- Status: Proposal

## Summary

This FLIP proposes to extend Flax's variable collections with a generic axis metadata API. 
The core of the API is an abstract base class that is recognized by lifting transformations that can add an axis (vmap, scan).
Users can extend the base class to keep track of per-axis metadata in a way that works with lifted transformations.

## Motivation

> Generally, there is no way in Flax to track metadata for variables across lifted transformations. Axis metadata is used to keep track of semantic information about axes into other (Flax independent) APIs.

Author: I think I've removed the worst offenders

For example, optimizers like AdaFactor can be configured on a per-axis level and partitioning APIs in JAX like xmap or pjit require per variable annotations to map effectiently to parallel hardware.

> Currently, there is an experimental [API](https://github.com/google/flax/blob/main/flax/linen/partitioning.py) supporting partitioning annotations with wrappers around lifted transforms that change axes (``nn.scan_with_axes``, ``nn.vmap+with_axes``) and a special APIs to create variables (``param_with_axes`` and ``variable_with_axes``).

Author: done

The experimental partitioning API stores the metadata in a seperate collection named "[collection]_axes".

The experimental API has a number of shortcomings that we like to solve:
1. The current API works for tracking PartitionSpecs but not for other types of metadata like optimizer annotations.
2. The implementation using an "xxx_axes" collection requires error-prone and non-composable string manipulation.
3. Special, partioning-aware variable creators and lifted transforms are required
4. The partioning API is hard to use with pre-existing Modules that aren't partioning aware.

## Proposal

To generalize metadata tracking and keep the specific metadata out of core Flax we propose the following abstract base class:

```python
class AxisMetadata(metaclass=abc.ABCMeta):

  @abc.abstractmethod
  def unbox(self) -> Any:
    pass

  @abc.abstractmethod

>   def add_axis(self, index: int, params: Dict[Any, Any]) -> TAxisMetadata:

Reviewer 3: why `TAxisMetadata` and not simply `AxisMetadata`?

Author: This is a subtitute for the missing Self type in Python see https://peps.python.org/pep-0673/

I did forget to add the self: TAxisMetadata annotation

    pass

  @abc.abstractmethod
  def remove_axis(self, index: int, params: Dict[Any, Any]) -> TAxisMetadata:
    pass
```

We call this type of class wrapping a value and keeping track of some additional data a **box**.
By defining an abstract base class for this box, the API does not need to be aware of the specifics of the metadata that is tracked.
This should make the API future proof and modular.

> The ``add_axis`` and ``remove_axis`` callback return an instance of their own type instead of mutating in-place.

Reviewer 3: I find the use of "callback" a bit confusing in this context. Why not simply "method"?

Author: agree

Typically, an implementation would be a ``flax.struct.PyTreeNode`` because the box should still be a valid JAX value and must therefore be handled by the PyTree API.
Calling ``jax.tree_map`` on a boxed value will simply map over the value in the box.
The lifted transforms that need to handle metadata will call ``jax.tree_map(..., is_leaf=lambda x: isinstance(x, AxisMetadata))`` to find the AxisMetadata instances within a PyTree.

Advantages of the boxing approach:
1. Boxing can be used outside of Flax and metadata is automatically "inherited". For example, the optimizerd state will
   have the same partitioning spec as the parameters, because the state is initialized using a ``jax.tree_map`` over the boxed parameters.
2. Boxes are composable.
3. Boxing avoids string manipulation and generally avoids having to handle additional auxilary collections like "param_axes" in the current
   partitioning API.
4. No need to lift metadata collections seperately.

> Disadvantages:

Reviewer 3: this also doubles the number of "levels" in the variable pytrees?
in particular, before the (un)boxing there was a 1:1 correspondence variables <-> module hierarchy
(which in turn might break some user code assumptions)

Author: Added

> 1. Handling boxed values requires the relatively new ``is_leaf=`` syntax which users might not be familiar with. Although users will

Reviewer 3: I wouldn't call this a disadvantage

   probably call Flax provided utils that handle the low-level tree_map calls in most cases.
3. Custom Pytree nodes have a small runtime overhead. It's hard to observe this in practise because JAX calls are async.

### Init syntax

Boxes can be created directly by the init function of a variable. Therefore, we propose to create metadata using higher-order initializers.
The main advantage of this is that we can decouple metadata handling compltely from the Module definition. Also, most Modules already overwrite
attributes to override the default initialzers so users can add metadata to existing Modules without requiring any code changes.

To illustrate this, let's consider a metadata class that keeps track of PartitionSpecs used by ``pjit``:

```python
class Partitioned(flax.struct.PyTreeNode, AxisMetadata):
  value: Any

>   named_axes: Tuple[Optional[str], ...]

Reviewer 3: `names` (or change below to `named_axes`)

>   named_axes: Tuple[Optional[str], ...]

Reviewer 3: wouldn't this metadata usually have `flax.struct.field(pytree_node=False)` ?

Author: Yes

Author: correct

Here we also defined a small utility called ``with_partitioning`` that we can use to wrap existing initialzers to add metadata: 

```python
# init kernel with lecun normal and split the output features over the data axis
partitioned_dense = nn.Dense(features, kernel_init=with_partitioning(nn.initializers.lecun_normal, (None, "data")))
```

### Unbox syntax

Metadata typically doesn't need to be handled by Modules directly. Therefore, we propose to make Modules agnostic to Metadata boxes by default.
The ``unbox`` method can be used to unpack a variable such that only the original JAX arrays remain. Users can manually call unbox but to make
sure Module classes don't have to call it everywhere we add an unbox keyword arg to variable returning APIs (e.g.: ``.param``, ``.variable``, ``.get_variable``).
The keyword arg ``unbox`` will default to ``True`` such that a Modules are metadata agnostic by default. This also means existing Modules will be backward compatible
with the new API.

```python
kernel = self.param("kernel", self.kernel_init, shape)  # No AxisMetadata instances
kernel_box = self.get_variable("param", "kernel", unbox=False)  # AxisMetadata boxes are preserved
```

> ### Lift syntax

Reviewer 3: to make the FLIP a bit more complete, it would be interesting to extend the `Partitioned` example a bit to show the entire lifecycle of the (un)boxing, i.e. lifting the `partitioned_dense`, showing the variable pytree etc.

Reviewer 3: Very nice illustration.

- FLIP PR: [#2435](https://github.com/google/flax/pull/2435)
- Status: Proposal

## Summary

This FLIP proposes to extend Flax's variable collections with a generic axis metadata API. 
The core of the API is an abstract base class that is recognized by lifting transformations that can add an axis (vmap, scan).
Users can extend the base class to keep track of per-axis metadata in a way that works with lifted transformations.

## Motivation

Generally, there is no way in Flax to track metadata for variables across lifted transformations. Axis metadata is used to keep track of semantic information about axes into other (Flax independent) APIs.
For example, optimizers like AdaFactor can be configured on a per-axis level and partitioning APIs in JAX like xmap or pjit require per variable annotations to map effectiently to parallel hardware.

Currently, there is an experimental [API](https://github.com/google/flax/blob/main/flax/linen/partitioning.py) supporting partitioning annotations with wrappers around lifted transforms that change axes (``nn.scan_with_axes``, ``nn.vmap_with_axes``) and a special APIs to create variables (``param_with_axes`` and ``variable_with_axes``).

> The experimental partitioning API stores the metadata in a seperate collection named "[collection]_axes".

Reviewer 4: seperate -> separate

The experimental API has a number of shortcomings that we like to solve:
1. The current API works for tracking PartitionSpecs but not for other types of metadata like optimizer annotations.
2. The implementation using an "xxx_axes" collection requires error-prone and non-composable string manipulation.
3. Special, partioning-aware variable creators and lifted transforms are required
4. The partioning API is hard to use with pre-existing Modules that aren't partioning aware.

## Proposal

To generalize metadata tracking and keep the specific metadata out of core Flax we propose the following abstract base class:

```python
class AxisMetadata(metaclass=abc.ABCMeta):

  @abc.abstractmethod
  def unbox(self) -> Any:
    pass

  @abc.abstractmethod

>   def add_axis(self: TAxisMetadata, index: int, params: Dict[Any, Any]) -> TAxisMetadata:

Reviewer 4: I think that adding the docstrings here already would be useful. 

Is add_axis called inside nn.scan? What does the user need to know about where and how it is called in order to use it properly.

Author: done

The lifted transforms that need to handle metadata will call ``jax.tree_map(..., is_leaf=lambda x: isinstance(x, AxisMetadata))`` to find the AxisMetadata instances within a PyTree.

Advantages of the boxing approach:
1. Boxing can be used outside of Flax and metadata is automatically "inherited". For example, the optimizerd state will
   have the same partitioning spec as the parameters, because the state is initialized using a ``jax.tree_map`` over the boxed parameters.
2. Boxes are composable.
3. Boxing avoids string manipulation and generally avoids having to handle additional auxilary collections like "param_axes" in the current
   partitioning API.
4. No need to lift metadata collections seperately.

Disadvantages:
1. Adding the boxes changes the PyTree hierarchy and introduces dataclasses within the otherwise plain, nested dict of variables.
3. Custom Pytree nodes have a small runtime overhead. It's hard to observe this in practise because JAX calls are async.

### Init syntax

Boxes can be created directly by the init function of a variable. Therefore, we propose to create metadata using higher-order initializers.

> The main advantage of this is that we can decouple metadata handling compltely from the Module definition. Also, most Modules already overwrite

Reviewer 4: completely

    pass

  @abc.abstractmethod
  def add_axis(self: TAxisMetadata, index: int, params: Dict[Any, Any]) -> TAxisMetadata:
    pass

  @abc.abstractmethod
  def remove_axis(self: TAxisMetadata, index: int, params: Dict[Any, Any]) -> TAxisMetadata:
    pass
```

We call this type of class wrapping a value and keeping track of some additional data a **box**.
By defining an abstract base class for this box, the API does not need to be aware of the specifics of the metadata that is tracked.
This should make the API future proof and modular.

The ``add_axis`` and ``remove_axis`` method return an instance of their own type instead of mutating in-place.
Typically, an implementation would be a ``flax.struct.PyTreeNode`` because the box should still be a valid JAX value and must therefore be handled by the PyTree API.
Calling ``jax.tree_map`` on a boxed value will simply map over the value in the box.
The lifted transforms that need to handle metadata will call ``jax.tree_map(..., is_leaf=lambda x: isinstance(x, AxisMetadata))`` to find the AxisMetadata instances within a PyTree.

Advantages of the boxing approach:

> 1. Boxing can be used outside of Flax and metadata is automatically "inherited". For example, the optimizerd state will

Reviewer 3: nit: "optimizer state"
